src/main/java/com/wanderersoftherift/wotr/init/WotrAttachments.java

```diff
@@ -56,12 +64,34 @@ public class WotrAttachments {
                             .build());
     public static final Supplier<AttachmentType<AbilitySlots>> ABILITY_SLOTS = ATTACHMENT_TYPES.register(
             "ability_slots",
-            () -> AttachmentType.builder(AbilitySlots::new).serialize(AbilitySlots.CODEC).copyOnDeath().build());
+            () -> AttachmentType.builder((holder) -> new AbilitySlots(holder, null))
+                    .serialize(new AttachmentSerializerFromDataCodec<AbilitySlots.Data, AbilitySlots>(
+                            AbilitySlots.Data.CODEC, AbilitySlots::new, AbilitySlots::data))
+                    .copyOnDeath()
+                    .build());
     public static final Supplier<AttachmentType<AttachedEffectData>> ATTACHED_EFFECTS = ATTACHMENT_TYPES.register(
             "attached_effects",
             () -> AttachmentType.builder(AttachedEffectData::new).serialize(AttachedEffectData.CODEC).build());
     public static final Supplier<AttachmentType<EffectDisplayData>> EFFECT_DISPLAY = ATTACHMENT_TYPES
             .register("effect_display", () -> AttachmentType.builder(() -> new EffectDisplayData()).build());
     public static final Supplier<AttachmentType<ManaData>> MANA = ATTACHMENT_TYPES.register("mana",
             () -> AttachmentType.builder(ManaData::new).serialize(ManaData.CODEC).build());
+
+    public static record AttachmentSerializerFromDataCodec<D, A>(Codec<D> codec,
```

Comment:
Could put this in the `codec` package, perhaps rename that package to `serialization`.